addon/components/bs-alert.js

```diff
@@ -163,6 +163,10 @@ export default class Alert extends Component {
    * @private
    */
   hide() {
+    if (this.hidden) {
+      return;
+    }
```

Comment:
Not a fan of adding a guard for an easy way out... I'd prefer if we can fix the underlying issue, but I wasn't sure what that issue is due to `_visible`, `@visible`, `hidden` seemingly tracking the same thing (I didn't investigate how these are different).

If we do go with adding the guard, I think I'd recommend adding an inline comment to explain its purpose.

Reply:
Yeah, I'll also prefer finding the root cause for this... maybe putting a break point into the `hide` method to see why it get's called twice? I am unsure why that is, given the [did-update](https://github.com/kaliber5/ember-bootstrap/blob/master/addon/components/bs-alert.hbs#L4) modifier should call it only once.

Reply:
I'm not sure yet why either. Yep, I can look into it further this week.

Reply:
I expected that the bug is caused by setting the property bound to `@visible` to `false` twice.

In the reproduction given in #1264 the `hideAlert` method is called twice. First by the event listener attached to the button and afterwards by `@onDismissed` event. `hideAlert` method sets `showAlert`, which is bound to `@visible` to `false` even if it's `false` before. This invalidates `_visible` property of the component, which triggers `{{did-update}}` modifier.

A simplified reproduction of that might look like this:

```js
this.visible = true;
await render(hbs`<BsAlert @visible={{this.visible}}>`);

// this is mocking the event listener on button
this.visible = false;

// await at least one runloop cause {{did-update}} modifier is not fired
// multiple times for the same runloop
await new Promise((resolve) => next(resolve));

// this is mocking the action bound to @onDismissed event
this.visible = false;
```

If I recall correctly the bug could not only be reproduced by clicking the button but also with the dismiss button rendered by the component itself. In that case it's a little bit more complex as two different states are changing: clicking the dismiss button rendered by alert component changes the internal state of `<BsAlert>`, `@onDismissed` event changes the state passed in from outside (`this.showAlert` bound to `@visible` argument). But as in the example before `_visible` property of `<BsAlert>` is invalidated twice, which triggers the `{{did-update}}` modifier twice.

To summarize: I assumed that setting property bound to `@visible` argument is required to reproduce the bug. I'm a little bit confused cause the tests added doesn't seem to do so. So maybe my assumption is wrong. Haven't verified.

If my assumption is right, adding a guard as you have done here, seems to be the correct fix in my opinion.

Reply:
Apologies for my slow progress with this pull request.

> // this is mocking the event listener on button
> this.visible = false;
>
> // await at least one runloop cause {{did-update}} modifier is not fired
> // multiple times for the same runloop
> await new Promise((resolve) => next(resolve));
>
> // this is mocking the action bound to @onDismissed event
> this.visible = false;

After the `await render()` call in my tests, I tried including these lines of code.

The line `await new Promise((resolve) => next(resolve));` didn't trigger the `this.dismissed` function that I passed to the component (i.e. `@onDismissed` was called 0 times).

If I continued to use `await settled()` instead, `@onDismissed` was called 1 time. Not sure yet what to make of this observation.